src/main/java/rx/internal/operators/OnSubscribeFromIterable.java

```diff
@@ -42,11 +43,24 @@ public OnSubscribeFromIterable(Iterable<? extends T> iterable) {
 
     @Override
     public void call(final Subscriber<? super T> o) {
-        final Iterator<? extends T> it = is.iterator();
-        if (!it.hasNext() && !o.isUnsubscribed())
+        final Iterator<? extends T> it;
+        boolean b;
+        
+        try {
+            it = is.iterator();
+            
+            b = it.hasNext();
+        } catch (Throwable ex) {
+            Exceptions.throwOrReport(ex, o);
+            return;
+        }
+            
+            
+        if (!b && !o.isUnsubscribed()) {
             o.onCompleted();
-        else 
+        } else { 
             o.setProducer(new IterableProducer<T>(o, it));
```

Comment:
Do we need to allocate and set the producer if unsubscribed? Seems like that check could enclose this whole if/else block.